com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/ProceduralVirtualTextureNode.cs

```diff
@@ -0,0 +1,105 @@
+using System.Collections.Generic;
+using UnityEditor.ShaderGraph.Drawing.Controls;
+using UnityEngine;
+using UnityEditor.Graphing;
+using UnityEditor.ShaderGraph.Internal;
+
+namespace UnityEditor.ShaderGraph
+{
+#if PROCEDURAL_VT_IN_GRAPH
+    [Title("Input", "Texture", "Procedural Virtual Texture")]
+    class ProceduralVirtualTextureNode : AbstractMaterialNode
+    {
+        public const int OutputSlotId = 0;
+
+        const string kOutputSlotName = "Out";
+
+        public ProceduralVirtualTextureNode()
+        {
+            UpdateNodeAfterDeserialization();
+            SetLayerCount(2);
+
+            vtProperty.displayName = "ProceduralVirtualTexture";
+            vtProperty.overrideReferenceName = "MyPVT";             // TODO : make unique
+            vtProperty.generatePropertyBlock = false;
+            vtProperty.value.procedural = true;
+
+            UpdateName();
+        }
+
+        void UpdateName()
+        {
+            name = "Procedural Virtual Texture: " + vtProperty.overrideReferenceName;
+        }
+
+        public sealed override void UpdateNodeAfterDeserialization()
+        {
+            AddSlot(new VirtualTextureMaterialSlot(OutputSlotId, kOutputSlotName, kOutputSlotName, SlotType.Output));
+            RemoveSlotsNameNotMatching(new[] { OutputSlotId });
+            SetLayerCount(layers);
+        }
+
+        [SerializeField]
+        private VirtualTextureShaderProperty vtProperty = new VirtualTextureShaderProperty();
+
+        void SetLayerCount(int layers)
+        {
+            var uniqueName = objectId;
+            vtProperty.value.layers.Clear();
+            layers = System.Math.Max(System.Math.Min(layers, SampleVirtualTextureNode.kMaxLayers), SampleVirtualTextureNode.kMinLayers);
+            for (int x = 0; x < layers; x++)
+            {
+                vtProperty.value.layers.Add(new SerializableVirtualTextureLayer("Layer" + x + "_" + uniqueName, "Layer" + x + "_" + uniqueName, null));
```

Comment:
We'll probably need to change the environment PVT layer names to this scheme when we want to enable using it for environment.

Reply:
There's actually no need to name the Layers at all..  all the setup in C# and Internal C++ is done by layer index, not by name.   The only reason there is a name at all is to provide the "public shader API" in shaderlab, where you can ask it to sample a layer by name instead of index.   ShaderGraph skips all of that and directly samples by index, so no need for the names at all really, except to satisfy the shader macros.

Reply:
I have to give the layers a unique name here so the shader macros don't create two functions with the same name.   (They declare "SampleVT_<LayerName>()" functions, which will collide if there are two layers with the same name)

Reply:
We could add a method to let the user name the layers, but all it would do is change unused bits of code in the generated shader.  Easier to read I guess, but no functional difference.  :)